nemo-physical/src/dictionary/hash_map_dictionary.rs

```diff
@@ -0,0 +1,413 @@
+        let mut page_num = self.cur_pages[buffer];
+        if self.pages[page_num].1.len() + len > PAGE_SIZE {
+            self.acquire_page_lock();
+            self.pages.push((buffer, String::with_capacity(PAGE_SIZE)));
+            page_num = self.pages.len() - 1;
+            self.cur_pages[buffer] = page_num;
+            self.release_page_lock();
+        }
+        let page_addr = self.pages[page_num].1.len();
+        self.pages[page_num].1.push_str(s);
+
+        StringRef::new(page_num * PAGE_SIZE + page_addr, s.len())
+    }
+
+    /// Returns a direct string slice reference for this data.
+    /// This is a pointer to global mutable data, and cannot be used safely.
+    fn get_str(&self, address: usize, length: usize) -> &str {
+        let page_num = address >> PAGE_ADDR_BITS;
+        let page_addr = address % PAGE_SIZE;
+        unsafe {
+            &self.pages[page_num]
+                .1
+                .get_unchecked(page_addr..page_addr + length)
+        }
+    }
+
+    /// Creates a reference to the given string without adding the string to the buffer.
+    fn get_tmp_string_ref(&mut self, buffer: usize, s: &str) -> StringRef {
+        self.tmp_strings[buffer].clear();
+        self.tmp_strings[buffer].push_str(s);
+        StringRef::new_tmp(buffer)
+    }
+
+    /// Returns the current contents of the temporary string.
+    fn get_tmp_string(&self, buffer: usize) -> &str {
+        self.tmp_strings[buffer].as_str()
+    }
+
+    /// Acquire the lock that we use for operations that add new pages or change
+    /// the assignment of pages to buffers in any way.
+    fn acquire_page_lock(&mut self) {
+        while self
+            .lock
+            .compare_exchange_weak(false, true, Ordering::Acquire, Ordering::Acquire)
+            .is_err()
+        {}
+    }
+
+    /// Release the lock.
+    fn release_page_lock(&mut self) {
+        self.lock.store(false, Ordering::Release);
+    }
+}
+
+/// Memory-optimized reference to a string in the dictionary.
+#[derive(Clone, Copy, Debug, Default)]
+struct StringRef {
+    /// The 64bits reference consists of 40bits that encode a starting address within
+    /// the buffer, and 24bits that encode the string length.
+    /// This limits the maximal buffer size to 1TB of string data, and the maximal length
+    /// of a single string to 16M bytes.
+    reference: u64,
+}
+impl StringRef {
+    /// Creates an object that refers to the current contents of the
+    /// buffer's temporary String.
+    fn new_tmp(buffer: usize) -> Self {
+        assert!(buffer < 1 << 24);
+        let u64buffer: u64 = buffer.try_into().unwrap();
+        StringRef {
+            reference: (u64::MAX << 24) + u64buffer,
+        }
+    }
+
+    /// Creates a reference to the specific string slice in the buffer.
+    /// It is not checked if that slice is allocated.
+    fn new(address: usize, len: usize) -> Self {
+        assert!(len < 1 << 24);
+        assert!(address < 1 << 40);
```

Comment:
Can we move these magic numbers to `const`s as well (also in other places in this `impl`)?